runtime/src/in_mem_accounts_index.rs

```diff
@@ -235,34 +235,96 @@ impl<T: IndexValue> InMemAccountsIndex<T> {
     ) -> RT {
         self.get_only_in_mem(pubkey, |entry| {
             if let Some(entry) = entry {
+                if entry.disk_unknown() {
+                    // disk_unknown is marked, load from the disk and merge with in-mem entry
+                    // we found the pubkey
+                    // but, we haven't checked to see if there is more on disk
+                    // so, we have to load from disk first, then merge with in-mem
+                    // right now we have a read lock on the in-mem idx, fwiw
+                    let disk_entry = self.load_account_entry_from_disk(pubkey);
+                    match disk_entry {
+                        Some(disk_entry) => {
+                            Self::merge_slot_lists(entry, disk_entry);
+                            entry.clear_disk_unknown();
+                        }
+                        None => {
+                            entry.clear_disk_unknown();
+                        }
+                    }
+                }
                 entry.set_age(self.storage.future_age_to_flush());
                 callback(Some(entry)).1
             } else {
                 // not in cache, look on disk
                 let stats = &self.stats();
                 let disk_entry = self.load_account_entry_from_disk(pubkey);
-                if disk_entry.is_none() {
```

Comment:
This code is moved to line 296. 
We keep disk_entry as an Option<> and handle Some<>, None cases separately for Occupied<> and Vacant<> map.entry() below.

Reply:
yes, but we have to get a write lock just to figure out that the item still isn't in the in-mem map. We already just checked that it wasn't in the in-mem map.

Reply:
Let me see if I understand your comments correct. If we know that `disk_entry.is_none()`, we don't need to check for in-mem `entry` and can return early. This will shorten the time to hold the write lock. This is correct for the original code without `disk_unknown`. However, with `disk_unkown`, we have to check the in-mem 'entry` when `disk_entry.is_one()` because we may need to clear `lazy_disk_load`.

Reply:
but we just checked in-mem that put us in the else clause and it wasn't there. Therefore, it cannot be `disk_unknown`. If we discover it isn't on disk `disk_entry.is_none()`, then it is not in memory and not on disk, so we can immediately return `None`

Reply:
Got it. Updated.